kpatch-build/kpatch-build

```diff
@@ -816,9 +855,30 @@ fi
 echo "Building original source"
 [[ -n "$OOT_MODULE" ]] || ./scripts/setlocalversion --save-scmversion || die
 unset KPATCH_GCC_TEMPDIR
+
+declare -a MAKEVARS
+if [ "$CONFIG_CC_IS_CLANG" -eq 1 ]; then
+	# CROSS_COMPILE does a bit more than just adding a prefix when used
+	# with clang, so we have to do this manually.
+
+	MAKEVARS+=("CC=$TOOLSDIR/kpatch-cc clang")
+
+	if [ -n "$LLVM" ]; then
+		MAKEVARS+=("LD=$TOOLSDIR/kpatch-cc ld.lld")
+	elif [ -n "$LD" ]; then
+		MAKEVARS+=("LD=$TOOLSDIR/kpatch-cc $LD")
+	else
+		# Defaulting to ld, still need to add kpatch-cc prefix
+		MAKEVARS+=("LD=$TOOLSDIR/kpatch-cc ld")
+	fi
+else
+	MAKEVARS+=("CROSS_COMPILE=$TOOLSDIR/kpatch-cc ")
```

Comment:
The only issue I found w/ CROSS_COMPILE for clang is that $kdir/Makefile tries to get the CLANG_FLAGS, --target, from the CROSS_COMPILE. CC and LD could have other names rather than gcc and clang (e.g., x86-linux-gnu-gcc). which makes CROSS_COMPILE provide better flexibility. Let's keep the CROSS_COMPILE here. As we know, kaptch-cc expects regex matching for tools as well.

Reply:
I agree that using CROSS_COMPILE would be more robust, however it sets `--target`, `--prefix` and `--gcc-toolchain`. And it also mangles them so here's how clang is called when using CROSS_COMPILE:

```
clang -Wp,-MMD,scripts/mod/.devicetable-offsets.s.d -nostdinc -isystem /usr/lib64/clang/10.0.1/include
-I./arch/x86/include -I./arch/x86/include/generated -I./include -I./arch/x86/include/uapi
-I./arch/x86/include/generated/uapi -I./include/uapi -I./include/generated/uapi -include ./include/linux/kconfig.h
-include ./include/linux/compiler_types.h -D__KERNEL__ -Qunused-arguments -Wall -Wundef
-Werror=strict-prototypes -Wno-trigraphs -fno-strict-aliasing -fno-common -fshort-wchar -fno-PIE
-Werror=implicit-function-declaration -Werror=implicit-int -Werror=return-type -Wno-format-security
-std=gnu89
--target=kpatch-gcc --prefix=/usr/libexec/kpatch/ /usr/bin/kpatch-gcc --gcc-toolchain=/usr/libexec/kpatch /usr
-no-integrated-as -Werror=unknown-warning-option -mno-sse -mno-mmx -mno-sse2 -mno-3dnow -m64
-mno-red-zone -mcmodel=kernel -Wno-sign-compare -fno-asynchronous-unwind-tables -O2
-Wframe-larger-than=1024 -fno-stack-protector -Wno-format-invalid-specifier -Wno-gnu -mno-global-merge
-pg -Wdeclaration-after-statement -Wvla -Wno-pointer-sign -fno-strict-overflow -fno-stack-check -Werror=date-time
-Wno-initializer-overrides -Wno-format -Wno-sign-compare -Wno-format-zero-length
-Wno-tautological-constant-out-of-range-compare -DKBUILD_MODFILE='"scripts/mod/devicetable-offsets"'
-DKBUILD_BASENAME='"devicetable_offsets"' -DKBUILD_MODNAME='"devicetable_offsets"' -fverbose-asm
-S -o scripts/mod/devicetable-offsets.s scripts/mod/devicetable-offsets.c
```

The interesing part is: `--target=kpatch-gcc --prefix=/usr/libexec/kpatch/ /usr/bin/kpatch-gcc --gcc-toolchain=/usr/libexec/kpatch /usr`

The build ultimately fails with `clang-10: error: no such file or directory: '/usr/bin/kpatch-gcc'`

So I don't think we can keep using CROSS_COMPILE without changes to how kbuild treats it in clang's case.

Reply:
Yup. We do have an issue w/ the Makefiles for kbuild w/ clang even though the fix in the Makefile is simple. The question is "can we (do we want to) update the Makefile not to parse CROSS_COMPILE?" Actually, CROSS_COMPILE is a prefix to CC, LD, ... So, those should be atomic rather than be parsed.

One concern on __not__ using CROSS_COMPILE is that CC and LD needs to be explicitly defined in kpatch-build. This requires an assumption where CC and LD needs to be on $PATH. In addition, if someone wants to use something like, x86_64-linux-gnu-gcc, explicit definition for CC and LD, could be an another issue.

Reply:
Our usage of CROSS_COMPILE="kpatch-cc " is a hack, so it didn't surprise me too much when it broke clang.  Anyway I'd be glad to get rid of it.  Also, eventually we may want to support the use of CROSS_COMPILE for **real** cross compiling.

If we allow the user to optionally specify CC and LD in the environment and/or kpatch-build command-line, does that resolve your concerns?

Reply:
Sadly, to remove the hack, the only way seems to optionally specify CC and LD.